rfcs/0108-nixos-containers.md

```diff
@@ -0,0 +1,467 @@
+a new implementation based on
+[`systemd-nspawn(5)`](https://man7.org/linux/man-pages/man5/systemd.nspawn.5.html) and incorporates
+[`systemd-networkd(8)`](https://man7.org/linux/man-pages/man8/systemd-networkd.service.8.html) for
+the networking stack rather than imperative networking while providing a reasonable upgrade path
+for existing installations.
+
+# Motivation
+[motivation]: #motivation
+
+The `nixos-container` feature originally appeared in `nixpkgs` in
+[2013](https://github.com/nixos/nixpkgs/commit/9ee30cd9b51c46cea7193993d006bb4301588001), at a time where `systemd` support was relatively new to NixOS.
+
+Back then, `systemd-nspawn` was
+[only designed as a development tool for systemd developers](https://lwn.net/Articles/572957/) and NixOS
+didn't [support networkd](https://github.com/NixOS/nixpkgs/commit/59f512ef7d2137586330f2cabffc41a70f4f0346).
+Due to those circumstances the entire feature was implemented
+in a fairly ad-hoc way. One of the most notable issues is the broken uplink during boot
+of a container:
+
+* Containers will be started in a template unit named [`container@.service`](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#Description). This
+  service [configures the network interfaces after the container has started](https://github.com/NixOS/nixpkgs/blob/2f96b9a7b4c083edf79374ceb9d61b5816648276/nixos/modules/virtualisation/nixos-containers.nix#L178-L229).
+
+* This means that even though the `network-online.target` is reached, no uplink is available
+  until the container is fully booted.
+
+  The implication is that a lot of services won't work as-is when installed into a container.
+  For instance, [oneshot](https://www.freedesktop.org/software/systemd/man/systemd.service.html#Type=) services
+  such as `nextcloud-setup.service` will hang if a database in e.g. a local network is used. Other
+  examples are `rspamd` or `clamav`.
+
+Additionally, we currently maintain a Perl script called `nixos-container.pl` which serves
+as the CLI frontend for the feature. This is not only an additional maintenance burden for us, but largely duplicates functionality already provided by [`machinectl(1)`](https://www.freedesktop.org/software/systemd/man/machinectl.html).
+
+The main reason why `machinectl` cannot be used as a complete replacement are
+[imperative containers](https://nixos.org/manual/nixos/stable/index.html#sec-imperative-containers)
+and state getting lost after the `container@<container-name>.service` unit
+has stopped since `.nspawn` units aren't used.
+
+In the following section the design of a replacement is proposed with these goals:
+
+* Use [`networkd`](https://www.freedesktop.org/software/systemd/man/systemd.network.html) as the networking stack since `systemd-nspawn` is part of the same project and
+  thus both components are designed to work together and issues like no uplink until the container is fully booted will be fixed.
```

Comment:
```suggestion
  thus both components are designed to work together and resolve issues like no uplink until the container is fully booted.
```

Nit: With "will be fixed" it isn't clear if this is *already* fixed in networkd+systemd-nspawn or if we expect that upstream will fix this issue because they work together. I think the new wording is more exact.

Reply:
Makes sense, applied, thanks!